modules/standard/Lists.chpl

```diff
@@ -0,0 +1,363 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  (Word smithing module text below).
+
+  ---
+
+  This module contains the implementation of a list datatype, intended to
+  replace the use of arrays to perform "vector-like" operations.
+
+  A List is a lightweight container intended to mimic Python's list datatype,
+  with fast O(log n)/O(1) (whatever we end up on) indexing  and an API close
+  in spirit to its Python counterpart.
+
+  The highly parallel nature of Chapel means that great care should be taken
+  when performing operations which may trigger a move of List elements.
+  Inserts and removals into the middle of a List are example operations which
+  may trigger a move, thus invalidating all references to elements of the
+  List held by tasks before the move. Appending an element to the end of a
+  List will never cause a move.
+
+  The following operations may trigger a move:
+    - insert
+    - remove
+    - reverse
+    - sort
+    - pop
+
+  ---
+
+  (End word smithing module text).
+
+  I know that @mppf in particular is very interested in limiting the number of
+  operations that move/displace List elements.
+
+  This pertains to the general issue of invalidating references to (and also
+  iterators over) List elements.
+
+  There is also the separate, more general issue of parallel safety. I'd
+  happily appreciate any/all recommendations for the "Chapeltastic" way of
+  making this List type parallel safe.
+
+  ---
+
+  I believe it is the intention of this new List datatype to replace all
+  "array-as-vec" operations on arrays. That means the following procedures
+  would no longer be supported (deprecated?, or removed?) on arrays:
+
+    - proc push_back(in val: this.eltType)
+    - proc push_back(vals: [])
+    - proc pop_back()
+    - proc push_front(in val: this.eltType)
+    - proc push_front(vals: [])
+    - proc pop_front()
+    - proc insert(pos: this.idxType, in val: this.eltType)
+    - proc insert(pos: this.idxType, vals: [])
+    - proc remove(pos: this.idxType)
+    - proc remove(pos: this.idxType, count: this.idxType)
+    - proc remove(pos: range(this.idxType, stridable = false))
+    - proc clear()
+
+  All of the methods given above change an array's domain (exactly what we
+  are trying to avoid, and the impetus for the List type).
+
+*/
+module Lists {
+
+  // Building blocks used to store "chunks" of List elements.
+  pragma "no doc"
+  class ListBlock {
+    type _etype;
+    var data: _ddata(_etype) = nil;
+  }
+
+  /*
+    A container intended to replace the use of arrays to perform "vector-like"
+    operations.
+
+    Is not currently parallel safe.
+
+    The following operations may trigger a move:
+      - insert
```

Comment:
It would be nice to link to the definitions of these functions elsewhere in the documentation.

Reply:
I don't know if I would repeat the information in the module documentation, but I think it is appropriate here

Reply:
To be sure, the module documentation is in _no way_ final at this point, and it contains a lot of the ramblings and musings that I put in the OP. I hope to have it sorted out after we get the API and semantics of List finalized.

But yes, that is a really good reminder, thanks!